changefeeccl: Projections and Filters in CDC.

miretskiy opened the pull request:
Add a variant of CHANGEFEED statement that allows specification
of predicates and projections.

```
CREATE CHANGEFEED [INTO 'sink'] [WITH opt=val, ...]
AS SELECT .... FROM t WHERE ...
```

This changefeed variant can target at most 1 table (and 1 column
family) at a time. The expressions used as the projections and
filters can be almost any supported expression with some restrictions:
  * Volatile functions not allowed.
  * Sub-selects not allowed.
  * Aggregate and window functions (i.e. functions operating over many
    rows) not allowed.
  * Some stable functions, notably functions which return MVCC
    timestamp, are overridden to return MVCC timestamp of the event.

In addition, some CDC specific functions are provided:
  * cdc_is_delete: returns true if the event is a deletion event.
  * cdc_prev: returns JSON representation of the previous row state.
  * cdc_updated_timestamp: returns event update timestamp (usually MVCC
    timestamp, but can be different if e.g. undergoing schema changes)
Additional CDC specific functions will be added in the follow on PRs.

Few examples:

* Emit all but the deletion events:
```
CREATE CHANGEFEED INTO 'kafka://'
AS SELECT * FROM table
WHERE NOT cdc_is_delete()
```

* Emit all events that modified `important_col` column:
```
CREATE CHANGEFEED INTO 'kafka://' WITH diff
AS SELECT *, cdc_prev() AS previous
FROM important_table
WHERE important_col != cdc_prev()->'important_col'
```

* Emit few colums, as well as computed expresions:
```
CREATE CHANGEFEED INTO 'kafka://' WITH diff
AS SELECT warehouseID, (totalItems - orderedItems) as itemsAvailable
FROM warehouse
WHERE region='US/east';
```

When filter expression is specified, changefeed will now consult
optimizer so that the set of spans scanned by changefeed can be
restricted based on the predicate.

For example, given the following table and a changefeed:
```
CREATE TABLE warehouse (
  region STRING,
  warehouseID int,
  ....
  PRIMARY KEY (region, warehouseID)
);

CREATE CHANGEFEED INTO 'kafka://' WITH diff
AS SELECT *
FROM warehouse
WHERE region='US/east';
```

The create changefeed will only scan table spans that contain `US/east`
region (and ignore all other table spans).

---

For foundational work, see:

- https://github.com/cockroachdb/cockroach/pull/81676
- https://github.com/cockroachdb/cockroach/pull/81249
- https://github.com/cockroachdb/cockroach/pull/80499

Addresses:
- https://github.com/cockroachdb/cockroach/issues/56949
- https://github.com/cockroachdb/cockroach/issues/31214


---

Release Notes (enterprise):
CHANGEFEED statement now supports general expressions -- predicates and projections.
Projections allow customers to emit only the data that they care about,
including computed columns, while predicates (i.e. filters) allow them
to restrict the data that's emitted only to those events that match the
filter.

```
CREATE CHANGEFEED INTO 'kafka://' AS SELECT * FROM t WHERE NOT cdc_is_delete()
```

Commit:
changefeedccl: Projections and Filters in CDC.

Add a variant of CHANGEFEED statement that allows specification
of predicates and projections.

```
CREATE CHANGEFEED [INTO 'sink'] [WITH opt=val, ...]
AS SELECT .... FROM t WHERE ...
```

This changefeed variant can target at most 1 table (and 1 column
family) at a time. The expressions used as the projections and
filters can be almost any supported expression with some restrictions:
  * Volatile functions not allowed.
  * Subselects not allowed.
  * Aggregate and window functions (i.e. functions operating over many
    rows) not allowed.
  * Some stable functions, notably functions which return MVCC
    timestamp, are overridden to return MVCC timestamp of the event.

In addition, some CDC specific functions are provided:
  * cdc_is_delete: returns true if the event is a deletion event.
  * cdc_prev: returns JSON representation of the previous row state.
  * cdc_updated_timestamp: returns event update timestamp (usually MVCC
    timestamp, but can be different if e.g. undergoing schema changes)
Additional CDC specific functions will be added in the follow on PRs.

Few examples:

* Emit all but the deletion events:
```
CREATE CHANGEFEED INTO 'kafka://'
AS SELECT * FROM table
WHERE NOT cdc_is_delete()
```

* Emit all events that modified `important_col` column:
```
CREATE CHANGEFEED INTO 'kafka://' WITH diff
AS SELECT *, cdc_prev() AS previous
FROM important_table
WHERE important_col != cdc_prev()->'important_col'
```

* Emit few colums, as well as computed expresions:
```
CREATE CHANGEFEED INTO 'kafka://' WITH diff
AS SELECT warehouseID, (totalItems - orderedItems) as itemsAvailable
FROM warehouse
WHERE region='US/east';
```

When filter expression is specified, changefeed will now consult
optimizer so that the set of spans scanned by changefeed can be
restricted based on the predicate.

For example, given the following table and a changefeed:
```
CREATE TABLE warehouse (
  region STRING,
  warehouseID int,
  ....
  PRIMARY KEY (region, warehouseID)
);

CREATE CHANGEFEED INTO 'kafka://' WITH diff
AS SELECT *
FROM warehouse
WHERE region='US/east';
```

The create changefeed will only scan table spans that contain `US/east`
region (and ignore all other table spans).

Release Notes (enterprise):
CHANGEFEED statement now supports general expressions: predicates and projections.
Projections allow customers to emit only the data that they care about,
including computed columns, while predicates (i.e. filters) allow them
to restrict the data that's emitted only to those events that match the
filter.

```
CREATE CHANGEFEED INTO 'kafka://' AS SELECT * FROM t WHERE NOT cdc_is_delete()
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82562)
<!-- Reviewable:end -->

miretskiy commented:
Note to the reviewers: I plan on expanding changefeed_test with more end2end tests.
In the meantime, want to start the review and see what TC has to say.

rytaft reviewed (commented):
Cool! I left a few nits. I think I reviewed the optimizer-related code, but if you could point me to specific files you want me to look at more closely that would be helpful.

> Reviewed 8 of 8 files at r1, 34 of 34 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-:-N420aUCD4J8SLPf6epR:bejbx35)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @stevendanna)

___
*[`-- commits` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N41xwlK0rUp9KJrQ4D7):*
now supports?
___
*[`pkg/ccl/changefeedccl/cdceval/constraint.go` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N41z9cRAgPl7jz2YyhW:-N41z9cRAgPl7jz2YyhX:b-pqczw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/cdceval/constraint.go#L25)):*
> ```Go
> 
> // ConstrainPrimaryIndexSpanByFilter attempts to constrain table primary index span via specified
> // filter.  Returns constrained span, and a possibly empty remaining filter that needs to be evaluated.
> ```

nit: comment should be <= 80 columns wide
___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N41zVKgE-6HnjUfLux4:-N41zVKgE-6HnjUfLux5:b-q507um) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L48)):*
> ```Go
> }
> 
> // NewEvaluatorForExpressions returns evaluator configured to process specified projection and filter expressions.
> ```

nit: wrap comment lines
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-3a35zR29AxW0xcH:-N42-3a35zR29AxW0xcI:b-azycf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/cdceval/validation.go#L85)):*
> ```Go
> 		return nil, pgerror.Newf(pgcode.InvalidParameterValue,
> 			"projections and filter cannot be used when running against multifamily table (table has %d families)",
> 			desc.NumFamilies())
> ```

I think you can relax this restriction in the future if the filters and projections only use columns from the target family
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 534 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-Pmv-CoybIUaY2AW:-N42-Pmv-CoybIUaY2AX:b-xykaxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L534)):*
> ```Go
> 
> func (tf *schemaFeed) validatePredicates(ctx context.Context, desc catalog.TableDescriptor) error {
> 	if len(tf.projection) == 0 {
> ```

did you mean predicate here?
___
*[`pkg/clusterversion/cockroach_versions.go` line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-gffAtTetXIdznJJ:-N42-gffAtTetXIdznJK:behvzj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/clusterversion/cockroach_versions.go#L368)):*
> ```Go
> 	// keys at the Pebble layer.
> 	EnablePebbleFormatVersionRangeKeys
> 	// EnablePredicateChangefeed indicates that changefeeds support predicates and projections.
> ```

nit: wrap comment
___
*[`pkg/clusterversion/cockroach_versions.go` line 369 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-k3gCfDsRx4_WrYc:-N42-k3gCfDsRx4_WrYd:bqc7n5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/clusterversion/cockroach_versions.go#L369)):*
> ```Go
> 	EnablePebbleFormatVersionRangeKeys
> 	// EnablePredicateChangefeed indicates that changefeeds support predicates and projections.
> 	EnablePredicateChangefeed
> ```

nit: maybe call this `EnablePredicateProjectionChangefeed`?
___
*[`pkg/sql/execinfrapb/processors_changefeeds.proto` line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-yyrDyoss1tpypWd:-N42-yyrDyoss1tpypWe:b-rd8h7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/sql/execinfrapb/processors_changefeeds.proto#L61)):*
> ```protobuf
> 
>   // Projection is the "select clause" for predicate changefeed.
>   optional Expression projection = 6 [(gogoproto.nullable) = false];
> ```

should this be repeated instead of optional?
___
*[`pkg/sql/opt/exec/execbuilder/builder.go` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N41vbG9EivpaFeSoDAf:-N41vbG9EivpaFeSoDAg:bt1oadl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c5546b14d4bb7d35328e13505525bc64c72612/pkg/sql/opt/exec/execbuilder/builder.go#L132)):*
> ```Go
> 	ContainsMutation bool
> 
> 	// wrapFunction returns resolvable function referenced for function with
> ```

nit: referenced -> reference
___
*[`pkg/ccl/changefeedccl/alter_changefeed_test.go` line 1285 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N420Kdo4ZwkzsQ9n1mY:-N420Kdo4ZwkzsQ9n1mZ:behvzj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/alter_changefeed_test.go#L1285)):*
> ```Go
> 	// and removed 'primary_key_filter'.  Since predicates and projections are no longer
> 	// a "string" option, alter statement implementation (and grammar) needs to be updated, and
> 	// this test modified and re-enabled.
> ```

nit: wrap comment


<!-- Sent from Reviewable.io -->

Commit:
sql: Return remaining filter when constraining spans.

Span constrainer can be used to constrain the span that needs
to be scanned via filter expression.  Prior to this change,
filter expression must have been fully satisfied by constrained
spans.  Now, the callers may request a remaining filter to be
returned if input filter cannot be fully satisfied.

In addition, augment execbuilder to use optional search path
in semantic context if it's available when resolving function
definitions.

Release Notes: None

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @rytaft, and @stevendanna)

___
*[`-- commits` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N41xwlK0rUp9KJrQ4D7):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

now supports?
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-3a35zR29AxW0xcH:-N42MsJCA2fh-zg52BH4:b-xvbl4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/cdceval/validation.go#L85)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think you can relax this restriction in the future if the filters and projections only use columns from the target family
</blockquote></details>

I agree; I spoke about this very problem before with @honoredb .  I left a TODO.  I think we should drop this restriction; but for now, keeping it.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 534 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-Pmv-CoybIUaY2AW:-N42NATD1n0-pdzUxYRE:bkrb729) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L534)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

did you mean predicate here?
</blockquote></details>

Nah.. Projection is right because you must have at least 1 expression in projection, while filter is optional.
I remaned this method to `validateProjectionAndFilter` to better reflect the purpose.
___
*[`pkg/clusterversion/cockroach_versions.go` line 369 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-k3gCfDsRx4_WrYc:-N42N_Bh9O8o9AcDKDFC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/clusterversion/cockroach_versions.go#L369)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: maybe call this `EnablePredicateProjectionChangefeed`?
</blockquote></details>

Done.
___
*[`pkg/sql/execinfrapb/processors_changefeeds.proto` line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N42-yyrDyoss1tpypWd:-N44BYvQBOyHsQWZKoPD:bg9k4vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/sql/execinfrapb/processors_changefeeds.proto#L61)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should this be repeated instead of optional?
</blockquote></details>

Nah.. I was actually putting entire list into that... I've been going back and forth on this... but basically
I decided to change sql grammar, and update this file approriately to have a tree.SelectClause
(and store entire select clause here, instead of splitting into projection/filter).
This way, I improve Go api (no more functions that take 2 strings where you can get the order wrong), and I also get ability to have "aliases".
___
*[`pkg/sql/opt/exec/execbuilder/builder.go` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N41vbG9EivpaFeSoDAf:-N42NhCXDUfoX8FVw-kT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c5546b14d4bb7d35328e13505525bc64c72612/pkg/sql/opt/exec/execbuilder/builder.go#L132)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: referenced -> reference
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/alter_changefeed_test.go` line 1285 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N420Kdo4ZwkzsQ9n1mY:-N42O4pZCvFxgkhxaOyo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8572ab098e9f5f8829c7b30d624135e77ddb9d5/pkg/ccl/changefeedccl/alter_changefeed_test.go#L1285)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: wrap comment
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (changes requested):
> Reviewed 1 of 8 files at r1, 8 of 36 files at r5, 2 of 37 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @rytaft, and @stevendanna)

___
*[`pkg/ccl/changefeedccl/changefeed_stmt.go` line 899 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4Xd1hTE_BVIK4xBRd8:-N4Xd1hTE_BVIK4xBRd9:bap628k) ([raw file](https://github.com/cockroachdb/cockroach/blob/8146460d4f14f0875bd54ff690cc3ce998c69ac7/pkg/ccl/changefeedccl/changefeed_stmt.go#L899)):*
> ```Go
> 	{
> 		if details.Select != "" {
> 			if len(details.TargetSpecifications) > 1 {
> ```

Suggest `len(details.TargetSpecifications) != 1` and `requires exactly one table` so that we'll catch it if this check happens before TargetSpecifications is populated.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 171 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4Xe8ICEvRUTuE2Kgoj:-N4Xe8ICEvRUTuE2Kgok:b9t4h4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8146460d4f14f0875bd54ff690cc3ce998c69ac7/pkg/ccl/changefeedccl/event_processing.go#L171)):*
> ```Go
> 		}
> 		updatedRow = projection
> 		prevRow = cdcevent.Row{}
> ```

Could you say more about why you're clearing this here? Maybe an alternative would be to call `c.evaluator.Projection(ctx, prevRow, mvccTimestamp, prevRow)` even though if the projection already references cdc_prev it's semantically a little messy.
___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 730 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4XgkRHBzUf826xVZ8P:-N4XgkRHBzUf826xVZ8Q:b-9q0lbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8146460d4f14f0875bd54ff690cc3ce998c69ac7/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L730)):*
> ```Go
> 	for _, c := range d.ResultColumns() {
> 		if c.Typ.UserDefined() {
> 			r.byName[c.Typ.Name()] = c.Typ
> ```

Annoying corner case here. You can't use enums across databases, but you can have enums with the same name from different schemas in the same table. 

```
root@localhost:26257/one> show create table both_switches;
   table_name   |                      create_statement
----------------+--------------------------------------------------------------
  both_switches | CREATE TABLE public.both_switches (
                |     a public.switch NULL,
                |     b alt.switch NULL,
                |     rowid INT8 NOT VISIBLE NOT NULL DEFAULT unique_rowid(),
                |     CONSTRAINT both_switches_pkey PRIMARY KEY (rowid ASC)
                | )
```

Name() looks like it returns the unqualified name (see types.go:1500) so this can collide. I think `c.TypeMeta.Name.FQName()` should do it.
___
*[`pkg/sql/parser/sql.y` line 3977 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4XjhOI0IvsQeyb0D0Y:-N4XjhOJ-L7N06wUcywX:b-j37v98) ([raw file](https://github.com/cockroachdb/cockroach/blob/8146460d4f14f0875bd54ff690cc3ce998c69ac7/pkg/sql/parser/sql.y#L3977)):*
> ```Yacc
>   }
> | CREATE CHANGEFEED /*$3=*/ opt_changefeed_sink /*$4=*/ opt_with_options
>   AS SELECT /*$7=*/target_list FROM /*$9=*/insert_target /*$10=*/opt_where_clause
> ```

Ooh, these arg number comments are handy.

`insert_target` feels wrong here, I think we should move away from coupling changefeed target syntax to other statements' target syntax even if they happen to be the same right now. `from_list` might be more appropriate or just add a changefeed_from_clause token.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 4016 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4XluwtEKlqiJ0hhNEI:-N4XluwtEKlqiJ0hhNEJ:blx1j59) ([raw file](https://github.com/cockroachdb/cockroach/blob/8146460d4f14f0875bd54ff690cc3ce998c69ac7/pkg/ccl/changefeedccl/changefeed_test.go#L4016)):*
> ```Go
> 	sqlDB := sqlutils.MakeSQLRunner(s.DB)
> 	// Create enum to ensure enum values displayed correctly in the summary.
> 	sqlDB.Exec(t, `CREATE TYPE status AS ENUM ('open', 'closed', 'inactive')`)
> ```

Couple of enum tests that might fail right now:
* Two enums with the same name, same database, different schemas
* Create a changefeed with `WHERE status='open'` and then `ALTER TYPE status RENAME VALUE 'open' to 'active'`.


<!-- Sent from Reviewable.io -->

miretskiy commented:
@rytaft  -- I would appreciate another 👁️  on optimizer related code.
In particular, I have made it possible to specify custom function resolver; and just minor fixes 
(uncovered during logic tests)..

@HonoreDB Added new tests: `TestChangefeedPredicateWithSchemaChange`, `TestChangefeedInvalidPredicate`,
`TestChangefeedConstrainsSpansBasedOnPredicate`

I've addressed most of your comments -- with exception of same named enum in different schemas.

> Suggest `len(details.TargetSpecifications) != 1` and `requires exactly one table` so that we'll catch it if this check happens before TargetSpecifications is populated.

Done.


> Could you say more about why you're clearing this here? Maybe an alternative would be to call `c.evaluator.Projection(ctx, prevRow, mvccTimestamp, prevRow)` even though if the projection already references cdc_prev it's semantically a little messy.

Added comments; I don't think it's correct to call projection on prev row.  I think the users should use cdc_prev;
I think we should refactor encoder at some point.  Just like I split up old `encodedRow` to be updated row and prev row, we should do something similar with the encoder where you encode row into a buffer; and then maybe encode previous row into the buffer, and then emit.  We should also get rid of "before"/"after" top level keys in the encoded data when using json encoder.


> Ooh, these arg number comments are handy.
> 
Thanks

> `insert_target` feels wrong here, I think we should move away from coupling changefeed target syntax to other statements' target syntax even if they happen to be the same right now. `from_list` might be more appropriate or just add a changefeed_from_clause token.

Okay; no more being "lazy".  For now, I added `changefeed_target_expr`, which I aliased to insert_target -- this is what I definitely need for now; we can change this if needed later.

> Create a changefeed with WHERE status='open' and then ALTER TYPE status RENAME VALUE 'open' to 'active'.
Added this test (see `TestChangefeedPredicateWithSchemaChange` test, `name:           "alter enum value",`) --  but it's disabled.  See comment why; Going to address this.

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @rytaft, and @stevendanna)

___
*[`pkg/ccl/changefeedccl/changefeed_stmt.go` line 919 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4hFdiuB6A0hHJ4wRW1:-N4hFdiuB6A0hHJ4wRW2:b-nqffs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3c51bd2e4b29f9f44d8a707e2faee4afb4e457/pkg/ccl/changefeedccl/changefeed_stmt.go#L919)):*
> ```Go
> 	}
> 
> 	if len(descriptors) != 1 {
> ```

I don't think it's possible to make this happen right now but in theory this check misses multiple targets when they're of the form "foo FAMILY f1, foo FAMILY f2".
___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval_test.go` line 154 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4hMCp30xQ6sO3a9KrJ:-N4hMCp30xQ6sO3a9KrK:bha5azd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3c51bd2e4b29f9f44d8a707e2faee4afb4e457/pkg/ccl/changefeedccl/cdceval/expr_eval_test.go#L154)):*
> ```Go
> 		},
> 		{
> 			testName:   "main/projection_aliased",
> ```

Is there a way to assert that the topic emitted to is bar, not foo? (It looks like that's your intent when there's an alias).


<!-- Sent from Reviewable.io -->

miretskiy commented:
> I don't think it's possible to make this happen right now but in theory this check misses multiple targets when they're of the form "foo FAMILY f1, foo FAMILY f2".

Yup; that's not possible because I removed this syntax option.  The reason for this is summarized here:
https://github.com/cockroachdb/cockroach/issues/82607


> Is there a way to assert that the topic emitted to is bar, not foo? (It looks like that's your intent when there's an alias).

Good point; not adding it in expr_eval_test (which really doesn't care about that stuff outside of being able to parse/eval expressions), and instead adding it to changefeed_test (`TestChangefeedPredicates`) -- which does care about that

Renabled test around enum value rename.

One outstanding issue, which may be left outstanding for now, is the fact that you can have same named enums in different schemas.

miretskiy commented:
@HonoreDB: I'm going to hold off on trying to fix same name enum across different schemas;
basically, even if I use fully qualified name, it won't be sufficient: enums could be used in expression that's not tied to table descriptor in any way (e.g. `select 'something'::other.enum`;
Pulling in full resolver is not trivial (requires txn for example) *and* would be very inefficient;
I will need to send follow on PRs to fix this.
In the meantime, I hope to merge this as is: if you disagree -- let me know; and if you agree, give me 👍

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just some nits

Btw, now that you are passing a full `SelectClause`, you could just run the full optimizer instead of the `ScalarBuilder`, which would enable a few more optimizations. Feel free to leave that for a future PR....

> Reviewed 3 of 38 files at r3, 1 of 38 files at r7, 2 of 38 files at r8, 29 of 36 files at r11, 7 of 7 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-:-N4sPyoQ8KKieybelyA-:bopr5q0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 174 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sIktV061GeTm_mBfJ:-N4sIktW5Cuv7adpbjsv:b57kbl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/changefeed_dist.go#L174)):*
> ```Go
> // fetchSpansForTable returns the set of spans for the specified table.
> // Usually, this is just the primary index span.
> // However, if "selectClause" is specified, the set of spans returned may be
> ```

I don't see `"selectClause"`. Maybe you mean `"details.Select"`?
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 176 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sJ3drEB6ns6RUOaEV:-N4sJ3drEB6ns6RUOaEW:b-qzey6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/changefeed_dist.go#L176)):*
> ```Go
> // However, if "selectClause" is specified, the set of spans returned may be
> // restricted to satisfy predicate in the select clause.  In that case,
> // possibly updated select clause returned representing the "remaining filter"
> ```

Is that the `string` return param? Maybe you could name the return params to make that more clear?
___
*[`pkg/ccl/changefeedccl/cdceval/constraint.go` line 72 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sKOe-ASArFgqfsW2e:-N4sKOe-ASArFgqfsW2f:b-kz0xn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdceval/constraint.go#L72)):*
> ```Go
> 	ed *cdcevent.EventDescriptor,
> ) ([]roachpb.Span, error) {
> 	// Predicate changefeed currently work on a single table only.
> ```

nit: currently works?
___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 102 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sL1T0CAqhlI-Vd9lR:-N4sL1T0CAqhlI-Vd9lS:b-q4ypkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L102)):*
> ```Go
> }
> 
> // configureProjection configures this evaluator to evaluate projection.
> ```

This comment doesn't match the function name
___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 124 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sLObK0wRdAw6C9BMc:-N4sLObK0wRdAw6C9BMd:bkplnip) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L124)):*
> ```Go
> 		}
> 		e.where = expr
> 		return nil
> ```

nit: this is unnecessary since you return nil below
___
*[`pkg/ccl/changefeedccl/cdceval/parse.go` line 30 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sMDpr-dXRZywyEyqo:-N4sMDpr-dXRZywyEyqp:b-b9lkd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdceval/parse.go#L30)):*
> ```Go
> 	}
> 
> 	return stmt.AST.(*tree.Select).Select.(*tree.SelectClause), nil
> ```

Can you be sure that this will always be a select clause? Is it possible this might panic?
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 55 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sMY4_-CnxYKfGTpe9:-N4sMY4_-CnxYKfGTpeA:bscql3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdceval/validation.go#L55)):*
> ```Go
> 	}
> 
> 	// Try to constrain spans by select clause.  We don't care about constrained spans here, but
> ```

nit: <= 80 columns wide
___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 99 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sMt38BXLYLMbVDJzQ:-N4sMt38BXLYLMbVDJzR:b-n0r3r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdceval/validation.go#L99)):*
> ```Go
> 	case jobspb.ChangefeedTargetSpecification_EACH_FAMILY:
> 		// TODO(yevgeniy): Relax this restriction; some predicates/projectsion
> 		// are entirely fine to use (e.g "*").
> ```

nit: "*" may be problematic, since the optimizer will expand it to every column in the table. You might want to allow selecting specific columns from the primary family, though.
___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 307 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sNMdD8lGp-V1MVsJg:-N4sNMdD8lGp-V1MVsJh:bk167sl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdcevent/event.go#L307)):*
> ```Go
> }
> 
> // TableDescriptor  returns underlying table descriptor.  This method is exposed
> ```

nit: extra space after TableDescriptor
___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 309 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sNUmJ56g4OOo-Sfg9:-N4sNUmJ56g4OOo-SfgA:b94qjuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/cdcevent/event.go#L309)):*
> ```Go
> // TableDescriptor  returns underlying table descriptor.  This method is exposed
> // to make it easier to integrate with the rest of descriptor APIs; prefer to use
> // higher level methods/structs (s.a. Metadata) instead.
> ```

nit: spell out "such as" or use "e.g."
___
*[`pkg/sql/parser/sql.y` line 3989 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sOXIUD-6ocNsW_JN2:-N4sOXIUD-6ocNsW_JN3:bskntyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/sql/parser/sql.y#L3989)):*
> ```Yacc
>     	Options: $4.kvOptions(),
>     	Targets: tree.ChangefeedTargets{target},
>     	Select:  &tree.SelectClause{
> ```

nit: some of these have tabs and some don't
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 6262 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4sPStO-3YJ3l8AOoyr:-N4sPStO-3YJ3l8AOoys:behvzj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a27742a699ffcbef4234fb7b4f69c315082f650/pkg/ccl/changefeedccl/changefeed_test.go#L6262)):*
> ```Go
> 
> 			// Execute afterAlterStmt immediately following alterStmt.
> 			// Sometimes, we need to e.g. insert new rows in order to observe changefeed error.
> ```

nit: wrap comment


<!-- Sent from Reviewable.io -->

miretskiy commented:
@rytaft TFTR.

> Btw, now that you are passing a full `SelectClause`, you could just run the full optimizer instead of the `ScalarBuilder`, which would enable a few more optimizations. Feel free to leave that for a future PR....

Yeah; still a bit too early for that.

> nit: "*" may be problematic, since the optimizer will expand it to every column in the table. You might want to allow selecting specific columns from the primary family, though.
This is one of the reasons why i'm not using full optimizer yet.

> Can you be sure that this will always be a select clause? Is it possible this might panic?
Shouldn't panic, but better safe than never; added appropriate checks.

All other comments addressed as well.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 36 of 36 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-:-N4yI89Z28s91skbPGUR:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)



<!-- Sent from Reviewable.io -->

HonoreDB reviewed (changes requested):
> Reviewed 2 of 8 files at r1, 3 of 38 files at r3, 1 of 38 files at r7, 36 of 36 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 730 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N4XgkRHBzUf826xVZ8P:-N55Zi0u8SXTM-KmGb0d:b-3pr2eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8146460d4f14f0875bd54ff690cc3ce998c69ac7/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L730)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Oh and also `ALTER type switch RENAME TO status` is a thing and arguably wouldn't be expected to break a running changefeed.
</blockquote></details>

I realize this is very unlikely to happen in the wild, but until enum name collision is addressed I think ideally we'd error out in ValidateTablesAndTargets when there are multiple enums with the same unqualified name rather than potentially emit incorrect data (e.g. imagine two enums `internal.role` and `external.role` where "admin" is a value for both but is the 0 value for one and the 1 value for the other.)
___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 137 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N55Z8a7Abp0wETurPt0:-N55Z8a7Abp0wETurPt1:brx46yc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8004b80981400f9d542268aa61ec8f4fbdfaeb9/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L137)):*
> ```Go
> 		}
> 
> 		if d.EqualsVersion(e.evaluator.EventDescriptor) {
> ```

Up to you, but maybe EqualsWithUDTCheck could return two bools for version matches and udts match so that we don't have to redo the version check.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 760 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N55auzZBQIwDNbJ8XNX:-N55auzZBQIwDNbJ8XNY:b-pgd1jz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8004b80981400f9d542268aa61ec8f4fbdfaeb9/pkg/ccl/changefeedccl/changefeedbase/options.go#L760)):*
> ```Go
> type Filters struct {
> 	WithDiff       bool
> 	IncludeVirtual bool
> ```

This is changing the intended semantics of changefeedbase.Filters, it was supposed to be specifically stuff that needs to be plumbed through to the rangefeed level (rangefeed.Filter). Feel free to make it be whatever's convenient, but please change the comment.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> This is changing the intended semantics of changefeedbase.Filters, it was supposed to be specifically stuff that needs to be plumbed through to the rangefeed level (rangefeed.Filter). Feel free to make it be whatever's convenient, but please change the comment.

Reverted this change -- copied IncludeVirtual method from your other pr instead.

> I realize this is very unlikely to happen in the wild, but until enum name collision is addressed I think ideally we'd error out in ValidateTablesAndTargets when there are multiple enums with the same unqualified name rather than potentially emit incorrect data (e.g. imagine two enums internal.role and external.role where "admin" is a value for both but is the 0 value for one and the 1 value for the other.)

Addressed.  Part of a big change (vs previous PR) is the addition of normalization step to the
changefeed expression.  Added new tests for normalization and validation logic (`validation_test.go`);
Expanded other tests to verify correct behavior.

> Up to you, but maybe EqualsWithUDTCheck could return two bools for version matches and udts match so that we don't have to redo the version check.
Done

@HonoreDB ptal.

HonoreDB reviewed (approved):
Some non-blocking comments, looks great! I do think ideally we'd disallow casting an enum to a string because of the corner case I mentioned in slack, but I'm not too worried about it--it's the sort of thing you could reasonably expect to break.

> Reviewed 14 of 44 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-:-N5Ks-EfCB6tk1Z42D8v:b-xflmf3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @HonoreDB, @miretskiy, @rytaft, and @stevendanna)

___
*[`pkg/ccl/changefeedccl/cdceval/validation.go` line 234 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N5KnxOd5sUUD8xvnPaY:-N5KnxOd5sUUD8xvnPaZ:bf1jrqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffd377df70bac971b5fba182ff19ec9f6a8911d4/pkg/ccl/changefeedccl/cdceval/validation.go#L234)):*
> ```Go
> 		if _, isAllowed := allowedOIDs[id]; !isAllowed {
> 			return pgerror.Newf(pgcode.FeatureNotSupported,
> 				"use of user defined types not references by target table is not supported")
> ```

typo: s/not references/not referenced.
Also, looks very important that the logic determining what gets stored in V.OIDs is the same as the logic in desc.UserDefinedTypeColumns, is there a way to make them both call the same function? (Maybe they do and I just haven't gotten there yet).
(Checked, they do both use the same logic).
___
*[`pkg/ccl/changefeedccl/cdctest/row.go` line 92 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N5KpFgv7ri8jor2uQqZ:-N5KpFgv7ri8jor2uQq_:b2yd1z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffd377df70bac971b5fba182ff19ec9f6a8911d4/pkg/ccl/changefeedccl/cdctest/row.go#L92)):*
> ```Go
> 		switch len(parts) {
> 		case 1:
> 			return "defaultdb", "public", parts[0]
> ```

aren't we usually in database `d` in tests?
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 6147 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/82562#-N5Kpn3eBHWSY-w-FSMS:-N5Kpn3eBHWSY-w-FSMT:b8hy2wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffd377df70bac971b5fba182ff19ec9f6a8911d4/pkg/ccl/changefeedccl/changefeed_test.go#L6147)):*
> ```Go
> 		// changefeed, user defined type values are replaced with their physical
> 		// representation.
> 		expectedExpr := normalizeCDCExpression(t, s.Server.ExecutorConfig(),
> ```

This makes sense but I think it means we need to disallow expressions where an enum is cast to a stringlike thing, as without a transaction they're in effect volatile.


<!-- Sent from Reviewable.io -->

miretskiy commented:
bors r+

jordanlewis commented:
This is super cool!

I was wondering, with a query like the following, why don't I get the projections I would expect?

```
root@localhost:26257/defaultdb> create changefeed with schema_change_policy='stop' as select a, b from a;

table,key,value
a,\x5b315d,\x7b226166746572223a207b2261223a20312c202262223a20327d7d
a,\x5b325d,\x7b226166746572223a207b2261223a20322c202262223a20327d7d
a,\x5b335d,\x7b226166746572223a207b2261223a20332c202262223a20327d7d
a,\x5b345d,\x7b226166746572223a207b2261223a20342c202262223a20327d7d
a,\x5b355d,\x7b226166746572223a207b2261223a20352c202262223a20327d7d
a,\x5b365d,\x7b226166746572223a207b2261223a20362c202262223a20327d7d
```

Instead of seeing the `table` `key` `value` format, I would expect to see projections `a` and `b` rendered as ordinary SQL data (integers in this case).

miretskiy commented:
I think you need cli changes @honoredb is trying to merge... Or specify
--format=bytea

On Fri, Aug 26, 2022, 11:41 PM Jordan Lewis ***@***.***>
wrote:

> This is super cool!
>
> I was wondering, with a query like the following, why don't I get the
> projections I would expect?
>
> ***@***.***:26257/defaultdb> create changefeed with schema_change_policy='stop' as select a, b from a;
>
> table,key,value
> a,\x5b315d,\x7b226166746572223a207b2261223a20312c202262223a20327d7d
> a,\x5b325d,\x7b226166746572223a207b2261223a20322c202262223a20327d7d
> a,\x5b335d,\x7b226166746572223a207b2261223a20332c202262223a20327d7d
> a,\x5b345d,\x7b226166746572223a207b2261223a20342c202262223a20327d7d
> a,\x5b355d,\x7b226166746572223a207b2261223a20352c202262223a20327d7d
> a,\x5b365d,\x7b226166746572223a207b2261223a20362c202262223a20327d7d
>
> Instead of seeing the table key value format, I would expect to see
> projections a and b rendered as ordinary SQL data (integers in this case).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/82562#issuecomment-1229113856>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVEGFZSIAPWSTI3O3J3V3GE45ANCNFSM5YESOD3A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

HonoreDB commented:
`SET SESSION bytea_output=escape` or https://github.com/cockroachdb/cockroach/pull/85181, yup. Not a problem with changefeed expressions, a problem with all sinkless changefeeds due to a recent CLI change.

jordanlewis commented:
Thanks. That improves things, but it wasn't really what I was after. I see this:

```
root@localhost:26257/defaultdb> create changefeed with schema_change_policy='stop' as select a, b from a;
table,key,value
a,[1],"{""after"": {""a"": 1, ""b"": 2}}"
a,[2],"{""after"": {""a"": 2, ""b"": 2}}"
a,[3],"{""after"": {""a"": 3, ""b"": 2}}"
a,[4],"{""after"": {""a"": 4, ""b"": 2}}"
a,[5],"{""after"": {""a"": 5, ""b"": 2}}"
a,[6],"{""after"": {""a"": 6, ""b"": 2}}"
```

With the new hotness (projections), I'd expect to see this:

```
a,b
1,2
2,2
3,2
```

etc. I'm sure this is different semantics to an ordinary changefeed, and perhaps we'd need a different way to ask for this output. It's also likely I'm wandering into an old debate so apologies if this has already been covered at length elsewhere (and please feel free to point me to old conversations and so on). But comparing to the kind of output you see from other streaming query systems like ksqldb, the table/key/value output is harder to consume and understand for a novice user.

Take a look at `EMIT CHANGES` output from ksqldb. As expected, it would "stream updates" to the given projections, returning the projected columns as native SQL, rather than wrapped in JSON.

![image](https://user-images.githubusercontent.com/43821/187115208-600e8668-b579-40bb-9d91-1de05cf6147b.png)

reference: https://www.confluent.io/blog/intro-to-ksqldb-sql-database-streaming/

miretskiy commented:
You are seeing "wrapped" envelope. We have an issue, and will address it
during stability, to not wrap by default when using expressions. I think
@honoredb has PR in flight for this already

On Sun, Aug 28, 2022, 11:09 PM Jordan Lewis ***@***.***>
wrote:

> Thanks. That improves things, but it wasn't really what I was after. I see
> this:
>
> ***@***.***:26257/defaultdb> create changefeed with schema_change_policy='stop' as select a, b from a;
> table,key,value
> a,[1],"{""after"": {""a"": 1, ""b"": 2}}"
> a,[2],"{""after"": {""a"": 2, ""b"": 2}}"
> a,[3],"{""after"": {""a"": 3, ""b"": 2}}"
> a,[4],"{""after"": {""a"": 4, ""b"": 2}}"
> a,[5],"{""after"": {""a"": 5, ""b"": 2}}"
> a,[6],"{""after"": {""a"": 6, ""b"": 2}}"
>
> With the new hotness (projections), I'd expect to see this:
>
> a,b
> 1,2
> 2,2
> 3,2
>
> etc. I'm sure this is different semantics to an ordinary changefeed, and
> perhaps we'd need a different way to ask for this output. It's also likely
> I'm wandering into an old debate so apologies if this has already been
> covered at length elsewhere (and please feel free to point me to old
> conversations and so on). But comparing to the kind of output you see from
> other streaming query systems like ksqldb, the table/key/value output is
> harder to consume and understand for a novice user.
>
> Take a look at EMIT CHANGES output from ksqldb. As expected, it would
> "stream updates" to the given projections, returning the projected columns
> as native SQL, rather than wrapped in JSON.
>
> [image: image]
> <https://user-images.githubusercontent.com/43821/187115208-600e8668-b579-40bb-9d91-1de05cf6147b.png>
>
> reference:
> https://www.confluent.io/blog/intro-to-ksqldb-sql-database-streaming/
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/82562#issuecomment-1229699184>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVEHZTH36ZYAQ5PBXN3V3QSXLANCNFSM5YESOD3A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

miretskiy commented:
Also, regarding native formatting instead of Json: this is something i want
to do for core feeds; probably next truest

On Mon, Aug 29, 2022, 8:47 AM Yevgeniy Miretskiy ***@***.***>
wrote:

> You are seeing "wrapped" envelope. We have an issue, and will address it
> during stability, to not wrap by default when using expressions. I think
> @honoredb has PR in flight for this already
>
> On Sun, Aug 28, 2022, 11:09 PM Jordan Lewis ***@***.***>
> wrote:
>
>> Thanks. That improves things, but it wasn't really what I was after. I
>> see this:
>>
>> ***@***.***:26257/defaultdb> create changefeed with schema_change_policy='stop' as select a, b from a;
>> table,key,value
>> a,[1],"{""after"": {""a"": 1, ""b"": 2}}"
>> a,[2],"{""after"": {""a"": 2, ""b"": 2}}"
>> a,[3],"{""after"": {""a"": 3, ""b"": 2}}"
>> a,[4],"{""after"": {""a"": 4, ""b"": 2}}"
>> a,[5],"{""after"": {""a"": 5, ""b"": 2}}"
>> a,[6],"{""after"": {""a"": 6, ""b"": 2}}"
>>
>> With the new hotness (projections), I'd expect to see this:
>>
>> a,b
>> 1,2
>> 2,2
>> 3,2
>>
>> etc. I'm sure this is different semantics to an ordinary changefeed, and
>> perhaps we'd need a different way to ask for this output. It's also likely
>> I'm wandering into an old debate so apologies if this has already been
>> covered at length elsewhere (and please feel free to point me to old
>> conversations and so on). But comparing to the kind of output you see from
>> other streaming query systems like ksqldb, the table/key/value output is
>> harder to consume and understand for a novice user.
>>
>> Take a look at EMIT CHANGES output from ksqldb. As expected, it would
>> "stream updates" to the given projections, returning the projected columns
>> as native SQL, rather than wrapped in JSON.
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/43821/187115208-600e8668-b579-40bb-9d91-1de05cf6147b.png>
>>
>> reference:
>> https://www.confluent.io/blog/intro-to-ksqldb-sql-database-streaming/
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cockroachdb/cockroach/pull/82562#issuecomment-1229699184>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ANA4FVEHZTH36ZYAQ5PBXN3V3QSXLANCNFSM5YESOD3A>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

jordanlewis commented:
Cool, thank you.
